lib/Logger/LogContext.cpp

```diff
@@ -0,0 +1,122 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2021 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "LogContext.h"
+
+#include "Basics/debugging.h"
+
+#include <string>
+#include <vector>
+
+using namespace arangodb;
+
+namespace {
+  thread_local LogContext localLogContext;
+}
+
+struct LogContext::Impl {
+  std::vector<std::pair<std::string, std::string>> values;
+};
+
+LogContext::ScopedValue::ScopedValue(std::string key, std::string value) {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  _newValue = value;
+#endif
+  std::tie(_idx, _oldValue) = LogContext::current().set(std::move(key), std::move(value));
+}
+
+LogContext::ScopedValue::~ScopedValue() {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  TRI_ASSERT(_idx < LogContext::current()._impl->values.size());
+  TRI_ASSERT(LogContext::current()._impl->values[_idx].second == _newValue);
+#endif
+  LogContext::current().restore(_idx, std::move(_oldValue));
+}
+
+LogContext::ScopedMerge::ScopedMerge(LogContext ctx) {
+  _values.reserve(ctx._impl->values.size());
+  for (auto const& v : ctx._impl->values) {
+    _values.emplace_back(v.first, v.second);
```

Comment:
Don't we want to call `set` instead? This depends on the intended semantics; currently if I set a value on a log context with a key that already exists, the value is replaced. However, when we merge, the key is inserted multiple times. I would be in favor of last or first values wins.

Reply:
Good point. I would be in favor or letting the last value win, i.e. simply overwrite existing values when they are stored again.